core/src/test/scala/unit/kafka/network/SocketServerTest.scala

```diff
@@ -184,6 +185,121 @@ class SocketServerTest extends JUnitSuite {
+
+      time.sleep(idleTimeMs + 1)
+      TestUtils.waitUntilTrue(() => openChannel(request2).isEmpty, "Failed to close idle channel")
+      assertTrue("Channel removed without processing staging receives", openOrClosingChannel(request2).nonEmpty)
+      processRequest(overrideServer.requestChannel, request2) // this triggers a failed send since channel has been closed
+      TestUtils.waitUntilTrue(() => openOrClosingChannel(request2).isEmpty, "Failed to remove channel with failed sends")
+      assertNull("Received request after failed send", overrideServer.requestChannel.receiveRequest(200))
+
+    } finally {
+      overrideServer.shutdown()
+      serverMetrics.close()
+    }
+  }
+
+  @Test
+  def testConnectionIdReuse() {
+    val idleTimeMs = 60000
+    val time = new MockTime()
+    props.put(KafkaConfig.ConnectionsMaxIdleMsProp, idleTimeMs.toString)
+    val serverMetrics = new Metrics
+    val overrideConnectionId = "127.0.0.1:1-127.0.0.1:2-0"
+    val overrideServer = new SocketServer(KafkaConfig.fromProps(props), serverMetrics, time, credentialProvider) {
+      override def newProcessor(id: Int, connectionQuotas: ConnectionQuotas, listenerName: ListenerName,
+                                protocol: SecurityProtocol): Processor = {
+        new Processor(id, time, config.socketRequestMaxBytes, requestChannel, connectionQuotas,
+          config.connectionsMaxIdleMs, listenerName, protocol, config, metrics, credentialProvider) {
+          override private[network] def connectionId(channel: SocketChannel): String = overrideConnectionId
+        }
+      }
+    }
+
+    def openChannel: Option[KafkaChannel] = overrideServer.processor(0).channel(overrideConnectionId)
+    def openOrClosingChannel: Option[KafkaChannel] = overrideServer.processor(0).openOrClosingChannel(overrideConnectionId)
+    def connectionCount = overrideServer.connectionCount(InetAddress.getByName("127.0.0.1"))
+
+    try {
+      overrideServer.startup()
+      val socket1 = connect(overrideServer)
+      TestUtils.waitUntilTrue(() => connectionCount == 1, "Failed to create channel")
+      val channel1 = openChannel
+
+      // Create new connection with same id when channel1 is still open and in Selector.channels
+      connect(overrideServer)
+      TestUtils.waitUntilTrue(() => connectionCount == 1, "Failed to close channel")
+      assertEquals(channel1, openChannel)
+
+      // Create new connection with same id when channel1 is closed with staged receives and in Selector.closingChannels
+      val serializedBytes = producerRequestBytes
+      (1 to 3).foreach(_ => sendRequest(socket1, serializedBytes))
+      val request = overrideServer.requestChannel.receiveRequest(2000)
+      time.sleep(idleTimeMs + 1)
+      TestUtils.waitUntilTrue(() => openChannel.isEmpty, "Idle channel not closed")
+      assertTrue("Channel removed without processing staging receives", openOrClosingChannel.nonEmpty)
+
+      connect(overrideServer)
```

Comment:
Maybe worth adding a comment explaining what we're doing here (like it was done for the other cases). Seems like we're testing that the closing channel will mean that the new connection is closed.